### PR TITLE
refactor: typescript jsdoc

### PR DESCRIPTION
**Pull request by evilebottnawi**
This PR contains a:

- [x] **bugfix**
- [ ] new **feature**
- [x] **code refactor**
- [x] **test update** 
- [ ] **typo fix**
- [x] **metadata update**

### Motivation / Use-Case

use JSDoc + fix some edge cases (let's add tests for edge cases in other PRs) + tests

### Breaking Changes

No

### Additional Info

No

### CODE REVIEW

File: types/ajv-keywords.d.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    | 'prohibited'
+    | 'deepProperties'
+    | 'deepRequired'
+    | 'uniqueItemProperties'
+    | 'regexp'
+    | 'formatMaximum'
+    | 'formatMinimum'
+    | 'formatExclusiveMaximum'
+    | 'formatExclusiveMinimum'
+    | 'dynamicDefaults';
```
Comment (vankop) [inline types/ajv-keywords.d.ts]: `allRequired`
`anyRequired`
`oneRequired`
also allowed

File: src/validate.js [hunk 2]
```diff
@@ -1,10 +1,33 @@
 import Ajv from 'ajv';
 import ajvKeywords from 'ajv-keywords';
 
-import absolutePathKeyword from './keywords/absolutePath';
+import addAbsolutePathKeyword from './keywords/absolutePath';
 
 import ValidationError from './ValidationError';
 
+/** @typedef {import("json-schema").JSONSchema4} JSONSchema4 */
```
Comment (vankop) [inline src/validate.js]: thats why I used in stylelint `stylelint` as package `import('stylelint').MyAwesomeType`. I think it does not matter much right now since there is no way to get `*.d.ts` files from jsdoc. Also I am curious about using reference here:
```
/// <reference path="../types/index.d.ts" />
```

File: src/validate.js [hunk 3]
```diff
@@ -1,10 +1,33 @@
...
+/** @typedef {import("json-schema").JSONSchema6} JSONSchema6 */
+/** @typedef {import("json-schema").JSONSchema7} JSONSchema7 */
+/** @typedef {import("ajv").ErrorObject} ErrorObject */
+
+/** @typedef {(JSONSchema4 | JSONSchema6 | JSONSchema7) & { patternRequired?: Array<string>; formatMinimum?: string; formatMaximum?: string; formatExclusiveMinimum?: boolean; formatExclusiveMaximum?: boolean; instanceof?: string | Array<string>; if?: Schema | boolean; then?: Schema | boolean; else?: Schema | boolean}} Schema */
```
Comment (vankop) [inline src/validate.js]: Unfortunately, it will not work as expected. For example for arrays https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/json-schema/index.d.ts#L104
TS will warn on custom keywords =(
Don't know how to provide types here better, maybe create custom JSON definitions with supporting template. Like:

in json schema definition
```
type JSONSchema4<T as {[k: string]: string | number | boolean | string[] | number[] | boolean[]}> = {
       items?: Array<JSONSchema4<T>>
} & T

then
```
type AdditionalKeywords = {formatMinimum?: string ...};
type Schema = JSONSchema4<AdditionalKeywords>
``

File: src/ValidationError.js [hunk 4]
```diff
@@ -640,18 +785,30 @@ class ValidationError extends Error {
...
   getSchemaPartText(schemaPart, additionalPath, needDot = false) {
-    if (additionalPath) {
+    if (!schemaPart) {
+      return '';
+    }
+
+    if (Array.isArray(additionalPath)) {
       for (let i = 0; i < additionalPath.length; i++) {
-        const inner = schemaPart[additionalPath[i]];
+        /** @type {Schema} */
```
Comment (vankop) [inline src/ValidationError.js]: ` /** @type {Schema | undefined} */`

on the next line there is guard for `isTruthy`

File: src/validate.js [hunk 5]
```diff
@@ -50,22 +83,35 @@ function validate(schema, options, configuration = {}) {
...
+/**
+ * @param {Array<ErrorObject>} errors
+ * @returns {Array<SchemaUtilErrorObject>}
+ */
 function filterErrors(errors) {
+  /** @type {Array<SchemaUtilErrorObject>} */
   let newErrors = [];
 
-  for (const error of errors) {
+  for (const error of /** @type {Array<SchemaUtilErrorObject>} */ (errors)) {
```
[the remark this answers is not in the document]
Comment (evilebottnawi) [reply src/validate.js]: I think that we can miss this here, we must finish this asap

File: src/validate.js (hunk 3 above)
Comment (evilebottnawi) [reply src/validate.js]: What about this solution https://github.com/epoberezkin/ajv-keywords/issues/53?

Comment (evilebottnawi) [reply src/validate.js]: I afraid here no good solutions, maybe we can use `/** typeof /** @type {{ patternRequired: string }} **/ (schema).patternRequired !== 'undefined' **/`, what do you think, also we can improve this in future

**Comment by evilebottnawi** [comment]
I think we can finish it now and continue to correct and improve in others PRs (yep, feel free to send a PR).

What here done:
- typescript@3.7 + types
- fixes some bugs and add additional check because ts show me problems :smile: 
- auto generate declaration from JSDoc
- small refactoring and improvements

File: src/validate.js (hunk 2 above)
Comment (evilebottnawi) [reply src/validate.js]: Now we can't use that, because our declarations generated from JSDoc

File: src/ValidationError.js (hunk 4 above)
Comment (evilebottnawi) [reply src/ValidationError.js]: :+1:

File: types/ajv-keywords.d.ts (hunk 1 above)
Comment (evilebottnawi) [reply types/ajv-keywords.d.ts]: Done :+1:

**Comment by vankop** [comment]
👍 I will take a look later what can be still improved
